noisy-simulator/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+    #[error("error when building operation: {0}")]
+    FailedToConstructOperation(String),
+    /// Failure when sampling instrument outcome.
+    #[error("numerical error: no outcome found when sampling instrument")]
+    FailedToSampleInstrumentOutcome,
+    /// Failure when sampling Kraus operators.
+    #[error("numerical error: no outcome found when sampling Kraus operators")]
+    FailedToSampleKrausOperators,
+    /// Provided an invalid state when creating or setting the state of the simulator.
+    #[error("provided an invalid state when creating or setting the state of the simulator: {0}")]
+    InvalidState(String),
+    /// `Matrix` ⋅ `Vector` multiplication mismatch.
+    #[error("matrix ⋅ vector multiplication mismatch; matrix is of dimension ({nrows}, {ncols}) but vector has {vec_dim} entries")]
+    MatrixVecDimensionMismatch {
+        /// Number of rows in the matrix.
+        nrows: usize,
+        /// Number of columns in the matrix.
+        ncols: usize,
+        /// Number of elements in the vector.
+        vec_dim: usize,
+    },
+    /// State is not normalized.
+    #[error("numerical error: trace should be between 0 and 1, but it is {0}")]
+    NotNormalized(f64),
+    /// A numerical error, such as a probability-0 event.
+    #[error("numerical error: probability-0 event")]
+    ProbabilityZeroEvent,
+    /// A qubit-id is greater than the number of qubits the simulation supports.
+    #[error("qubit id out of bounds: {0}")]
+    QubitIdOutOfBounds(usize),
+    /// Failure when building a `StateVector` from raw data.
+    #[error("error when building `StateVector` from raw_data: {0}")]
+    StateVectorTryFromError(String),
+    /// Trace is not real
+    #[error("state trace should be real since it represents a probability, but its imaginary part is: {0}")]
+    TraceIsNotReal(f64),
+}
+
+impl Error {
+    const fn is_unrecoverable(&self) -> bool {
+        matches!(
+            self,
+            Error::ProbabilityZeroEvent
+                | Error::FailedToSampleInstrumentOutcome
+                | Error::FailedToSampleKrausOperators
+        )
+    }
+}
+
+impl From<&Error> for Error {
+    fn from(value: &Error) -> Self {
+        value.clone()
+    }
+}
+
+impl From<&mut Error> for Error {
+    fn from(value: &mut Error) -> Self {
+        value.clone()
+    }
+}
+
+/// Handles errors in the simulator.
+/// If an error is unrecoverable, it will set the state of the simulator to that error,
+/// invalidating any further evolution of the quantum system.
+macro_rules! handle_error {
+    ($self:expr, $err:expr) => {
+        if $err.is_unrecoverable() {
+            $self.state = Err($err.clone());
+        }
+        return Err($err)
+    };
+}
+
+pub(crate) use handle_error;
+
+pub use {
+    density_matrix_simulator::{DensityMatrix, DensityMatrixSimulator},
+    instrument::Instrument,
+    operation::Operation,
+    state_vector_simulator::{StateVector, StateVectorSimulator},
+};
```

Comment:
These should be at the top

Reply:
Moved to the top, `pub(crate) handle_error` needs to stay there because it needs to be after the macro definition. Else it does not compile.